File: coding_challenge/Python/28_remove_dup_node_linkedlist.py
# With set

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def remove_duplicates(head):
    if not head:
        return

    seen = set()
    current = head
    prev = None

    while current:
        if current.data in seen:
            prev.next = current.next  # Duplicate found, skip the node
        else:
            seen.add(current.data)  # First occurrence, add to set
            prev = current
        current = current.next

def print_list(head):
    current = head
    while current:
        print(current.data, end=" ")
        current = current.next
    print()

# Example usage
head = Node(10)
head.next = Node(12)
head.next.next = Node(11)
head.next.next.next = Node(12)
head.next.next.next.next = Node(10)

print("Original List:")
print_list(head)

remove_duplicates(head)

print("List after removing duplicates:")
print_list(head)


# def remove_duplicates_no_buffer(head):
#     if not head:
#         return

#     current = head

#     while current:
#         runner = current
#         while runner.next:
#             if runner.next.data == current.data:
#                 runner.next = runner.next.next  # Remove duplicate
#             else:
#                 runner = runner.next
#         current = current.next

File: coding_challenge/C++/28_remove_dup_node_linkedlist.cpp
#include <iostream>
#include <unordered_set>

struct Node {
    int data;
    Node* next;
    Node(int d) : data(d), next(nullptr) {}
};

void removeDuplicates(Node* head) {
    if (!head) return;

    std::unordered_set<int> seen;
    Node* current = head;
    Node* prev = nullptr;

    while (current) {
        if (seen.find(current->data) != seen.end()) {
            // Duplicate found, remove it
            prev->next = current->next;
            delete current;
        } else {
            // First occurrence, add to the set
            seen.insert(current->data);
            prev = current;
        }
        current = prev->next;
    }
}

// Helper function to print the linked list
void printList(Node* head) {
    while (head) {
        std::cout << head->data << " ";
        head = head->next;
    }
    std::cout << std::endl;
}

int main() {
    Node* head = new Node(10);
    head->next = new Node(12);
    head->next->next = new Node(11);
    head->next->next->next = new Node(12);
    head->next->next->next->next = new Node(10);

    std::cout << "Original List: ";
    printList(head);

    removeDuplicates(head);

    std::cout << "List after removing duplicates: ";
    printList(head);

    return 0;
}



// void removeDuplicatesNoBuffer(Node* head) {
//     if (!head) return;

//     Node* current = head;

//     while (current) {
//         Node* runner = current;
//         while (runner->next) {
//             if (runner->next->data == current->data) {
//                 // Duplicate found, remove it
//                 Node* temp = runner->next;
//                 runner->next = runner->next->next;
//                 delete temp;
//             } else {
//                 runner = runner->next;
//             }
//         }
//         current = current->next;
//     }
// }